keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md

```diff
@@ -0,0 +1,220 @@
+* Build user documentation for out-of-tree credential providers
+* Migrate existing in-tree credential providers to new credential provider interface
+* Remove in-tree credential provider code from kuberentes core
+
+### Non-Goals
+
+* Broad removal of cloud SDK usage falls under the [KEP for removing in-tree providers](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/2019-01-25-removing-in-tree-providers.md).
+
+## Proposal
+
+### External Credential Provider
+
+An executable capable of providing container registry credentials will be installed on each node. This binary will be executed by the kubelet to obtain container registry credentials in a format compatible with container runtimes. Credential responses may be cached within the kubelet.
+
+This Architecture is similar to the approach taken by CNI and is a well understood deployment pattern.  The API types are modelled after the ExecConfig and ExecCredential in client-go which define exec based credential retrieval for similar use cases.
+
+A RegistryCredentialProvider API type will be added to Kubernetes:
+
+```go
+// RegistryCredentialProvider is used by the kubelet container runtime to match image URIs with
+// exec-based credential provider plugins that provide container registry credentials.
+type RegistryCredentialProvider struct {
+    metav1.TypeMeta `json:",inline"`
+
+    // Spec holds configuration for the credential provider
+    Spec RegistryCredentialProviderSpec `json:"spec,omitempty"`
+}
+
+type RegistryCredentialProviderSpec struct {
+    // ImageMatchRegex is used to match against the Image URI to determine which images to
+    // provide credentials for
+    ImageMatchRegex string
+
+    // Exec specifies a custom exec-based authentication plugin.  This type is defined in
+    // https://github.com/kubernetes/client-go/blob/62f256057db7571c5ed1aba47eea291f72dd557a/tools/clientcmd/api/types.go#L184
+    Exec clientcmd.ExecConfig
+}
+```
+
+The existing `ExecCredentialSpec` and `ExecCredentialStatus` will add support for registry credentials.  In the in-tree implementation, the docker keyring,
+which has N credential providers, returns an `[]AuthConfig` on a Lookup(image string) call.
+
+```go
+// ExecCredentialSpec holds request and runtime specific information provided by
+// the transport.
+type ExecCredentialSpec struct {
+    // +optional
+    Response *Response `json:"response,omitempty"`
+
+    // +optional
+    Interactive bool `json:"interactive,omitempty"`
+
+    // Adding field:
+    // Image is used when passed to registry credential providers to as part of an image pull
+    // +optional
+    Image string `json:"image,omitempty"`
+}
+
+// ExecCredentialStatus holds credentials for kube runtime to use for image pulls.
+type ExecCredentialStatus struct {
+    // +optional
+    ExpirationTimestamp *metav1.Time `json:"expirationTimestamp,omitempty"`
+    Token string `json:"token,omitempty"`
```

Comment:
How is this used? Can you clarify the difference with the Identity & Registry token?

Reply:
Because I was adding onto the exec based credential structs, this was present but would have been unused.  After forking the structs I think it makes more sense now.